### DISCUSSION TITLE
C# bindings

### DISCUSSION THREAD

**Discussion by weianweigan** [elalish/manifold#900] [Show and tell]
# c# bindings :computer:

Hello everyone, Manifold is a awsome library. 
I am a C# developer and often develop plugins for CAD software. It would be fantastic if ManifoldCAD could be integrated into these design and manufacturing software. For industrial software like Solidworks, AutoCAD, Revit, etc., C# is a mainstream language for plugin development. Therefore, I am creating C# bindings and released an alpha version of a NuGet package for seamless integration into these 3D CAD software.
Hope this is helpful for C# developers.
This is the [repository address](https://github.com/weianweigan/manifold-csharp) and [introduction file](https://github.com/weianweigan/manifold-csharp/blob/dev-csharp/bindings/csharp/Readme.md) that I forked.

# Integrating with other 3D software. :gear:

I have already tried integrating Manifold with [Autodesk Revit](https://www.autodesk.com/sg/products/revit/overview), and it works fine. If you also want to integrate Manifold into 3D CAD software, you can give it a try.

# c# notebooks :notebook:

I attempted to use the C# version of [notebook](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.dotnet-interactive-vscode) to test and validate the manifold. I found that it is feasible, so I created a NuGet package. You can try using the C# notebook in VS Code to generate some models for quick validation.

[c# notebooks for manifold](https://github.com/weianweigan/manifold-csharp/blob/dev-csharp/bindings/csharp/notebooks/01.GetStarted.ipynb)

> It seems that 3D models cannot be viewed on GitHub, but they can be previewed in VS Code after they are generated.

<img src="https://github.com/user-attachments/assets/64a0d906-3290-4276-8f0f-3af0b8fe1f7a" height="500">

**Comment by elalish** [comment]
Awesome, thanks for sharing! We should add this to our README too, right next to the Java bindings.

**Reply by weianweigan** [reply]
Great! c# developer can find it quickly.

**Reply by edumarzolla** [reply]
Please help me!!!!

I'm already trying to solve during the last 48 hours on your wonferful ManifoldNET.Meshio nuget lib...
Everything was working fine since you implemented the ManifoldNET... probably after a windows update, the code stops here:

                Manifold cylinder1= Manifold.Cylinder((float)(10), (float)(5/ 2), (float)(5/ 2), 360, true);
                Manifold cylinder2= Manifold.Cylinder((float)(11), (float)(3 / 2), (float)(3 / 2), 360, true);
                Manifold result = Manifold.BooleanOperation(cylinder1, cylinder2, BoolOperationType.Subtract);   (the code stoped here at the Boolean operation line, with the following error)

System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

'Posibras-Came.exe' (CoreCLR: clrhost): Loaded 'C:\Users\admin\My Drive\aPosibras (1)\Calculo de came\Posibras_C\Posibras_C_1.1.79\Posibras-Nuget\Posibras-Came\bin\Debug\net8.0-windows7.0\ManifoldNET.dll'. 
An unhandled exception of type 'System.AccessViolationException' occurred in ManifoldNET.dll
Attempted to read or write protected memory. This is often an indication that other memory is corrupt.

'Posibras-Came.exe' (CoreCLR: clrhost): Loaded 'C:\Program Files\Microsoft Visual Studio\2022\Community\Common7\IDE\PrivateAssemblies\Runtime\Microsoft.VisualStudio.Debugger.Runtime.NetCoreApp.dll'. 
The program '[19940] Posibras-Came.exe' has exited with code 4294967295 (0xffffffff).
                

The exactly same code is running OK in other PC, with same Windows11 sysop, same VS2022 with the same installation. I already tryed to format the PC and install abrand new Windows11 with new VS2022 C#

I am getting crazy because I cannot understand what is happening...

Could you please help me with this crazy issue?

Thanks!!

Eduardo
+5511 966974533 (whatsapp)

**Reply by elalish** [reply]
I have no idea, nor a windows machine to repro on. Hopefully @weianweigan has some ideas. I'm guessing this goes through our C bindings? Perhaps a problem there - C memory management is easy to get wrong.

**Reply by weianweigan** [reply]
@edumarzolla  I didn't reproduce this error using `nuget: ManifoldNET.MeshIO 1.0.7-alpha`. Could you provide the version of the nuget package you are using?
Additionally, you should pay attention to the issue of floating-point type conversion.
![image](https://github.com/user-attachments/assets/e69927c7-03c6-40db-abc8-b9f565c2565c)

**Reply by pca006132** [reply]
probably @edumarzolla

**Reply by weianweigan** [reply]
@pca006132 Thanks！

**Comment by edumarzolla** [comment]
Fantastic!!!!!
How to keep nuget updated with manifold new evolutions?
Thanks

**Reply by weianweigan** [reply]
I added [a GitHub action](https://github.com/weianweigan/manifold-csharp/blob/dev-csharp/.github/workflows/nuget.yml) to push the NuGet package. The C# code is located in the /bindings/csharp directory. If Manifold didn't make any breaking changes, simply syncing the forked repos with the Manifold repos work, and then trigger GitHub workflow should works.

**Comment by edumarzolla** [comment]
If I understand correctly:
- ManifoldNET is the Manifold without Meshio
- ManifoldNET.Meshio is the Manifold with Meshio
- ManifoldNET.Noteooks is the Manifold with Meshio and visualization screen

Do you have an example of  ManifoldNET.Notebooks C# code?

Thanks!

Eduardo
+5511 966974533 (whatsapp)

Em sex., 16 de ago. de 2024 às 04:00, dududu ***@***.***>
escreveu:

**Reply by weianweigan** [reply]
It's right. [Notebooks sample here](https://github.com/weianweigan/manifold-csharp/tree/dev-csharp/bindings/csharp/notebooks). If you want to use it in a WinForms application, [ManifoldNET.MeshIO](https://www.nuget.org/packages/ManifoldNET.MeshIO) is recommended.

About notebook, you need vs code and [polyglot notebooks extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.dotnet-interactive-vscode) installed.

**Reply by edumarzolla** [reply]
Great!
I developed a code in C# that creates a solid using compiled Manifold DLL.
How to visualize the solid in a WindowsForm in c# (Visual Studio 2022, not VS code)
Thanks

**Reply by weianweigan** [reply]
@edumarzolla  Try [helix-toolit](https://github.com/helix-toolkit/helix-toolkit). It may be necessary to place a WPF control inside a WinForm. Webview may be another option.

**Reply by weianweigan** [reply]
helix-toolkit example which load glb file [here](https://github.com/helix-toolkit/helix-toolkit/tree/develop/Source/Examples/WPF/ExampleBrowser/Examples/ModelViewer).

**Reply by edumarzolla** [reply]
Hi,
When you created the nuget pack, have you compiled the dll with TBB (parallel processing) standards?
Asking that because when I running multiple boolean operations, it is take long time, and the process graph (task manager) is not showing all CPUs processing

![image](https://github.com/user-attachments/assets/34e1c5ee-d8af-4a88-ac1d-5f3076411cfd)

Thanks

**Comment by edumarzolla** [comment]
I replaced the manifoldc.dll file from the nuget folder with a
manifoldc.dll that I compiled with TBB, and the result was 10 times faster.
I made a backup before. Compare the size of files...
[image: image.png]


Em sex., 16 de ago. de 2024 às 12:14, dududu ***@***.***>
escreveu:

**Reply by weianweigan** [reply]
@edumarzolla Thank you for pointing that out. I compiled Manifold in GitHub Action, but TBB is not enabled yet.

**Comment by edumarzolla** [comment]
Please see the attached document that I created for compiling it with TBB.


Em qui., 22 de ago. de 2024 às 22:55, dududu ***@***.***>
escreveu:

**Reply by weianweigan** [reply]
@edumarzolla I didn't see the attached document, possibly because you replied directly from the email.

**Reply by edumarzolla** [reply]
@weianweigan 
[How to compile Manifold DLL in VS2022.docx](https://github.com/user-attachments/files/16736802/How.to.compile.Manifold.DLL.in.VS2022.docx)

**Reply by weianweigan** [reply]
@edumarzolla Thanks! I compiled with tbb option on and released new version [1.0.7-alpha](https://www.nuget.org/packages/ManifoldNET.MeshIO/1.0.7-alpha). It contains tbb12.dll in runtimes folder. I run all test and passed, it seems faster.

**Comment by edumarzolla** [comment]
Wonderful!!! I will test it tomorrow, now is Sunday 11pm in São
paulo/Brazil.
I miss a visualizer for c# windows form apps, for me, if your lib could add
a toolbox for visualizing,  it will get still better!!

Thanks

**Reply by weianweigan** [reply]
@edumarzolla  Great! Hope the test goes smoothly. NanJing/China is now 10:30 am. I had a [library](https://github.com/weianweigan/ModelViewerWPF) for Visualization based on wpf and it had an example about how to intergate with manifold-csharp. It also can be used in winform by [elementhost](https://learn.microsoft.com/en-us/dotnet/desktop/winforms/advanced/walkthrough-creating-new-wpf-content-on-windows-forms-at-design-time?view=netframeworkdesktop-4.8).
![image](https://github.com/user-attachments/assets/371a7f82-1ec3-4a97-b0b8-939dce36101a)

**Comment by reinux** [comment]
This is awesome. I was wondering about exactly this just last month. The timing couldn't be better. Thanks!

I have an idea for a pet project translating OpenSCAD files to F#.

**Reply by elalish** [reply]
Nice, and in case it's helpful: https://github.com/cookiecad/Openscad-to-JSCAD-converter/issues/10

**Reply by reinux** [reply]
That is indeed helpful. Thanks!

**Comment by elalish** [comment]
@weianweigan @edumarzolla @reinux I'm taking a wild guess some of you may be Windows users? We just hit a strange error on our Windows CI, tracked by #910 - the problem is neither @pca006132 nor I have a Windows box and the error is completely obtuse. You should be able to repro if you pull master and run `manifold_test.exe --gtest_filter=Polygon.Zebra` or `Zebra3`. If any of you can repro, any info you can give us to help narrow down the problem would be a huge help. Thanks!

**Reply by reinux** [reply]
Unfortunately I switched all my machines to Linux recently. Sorry!

**Reply by weianweigan** [reply]
@reinux Please note that the C# bindings for Manifold have not yet included binaries for Linux and macOS. Currently, I have only compiled the Windows version.

**Reply by reinux** [reply]
@weianweigan I was actually able to get it working with fairly little effort just using the flake.nix from @elalish 's original repo, save for a minor nuance which I noted in https://github.com/elalish/manifold/issues/909 . I think on Ubuntu and others, it should suffice to use the package manager.

Otherwise, the only wrinkle was with loading extension.dib in an F# notebook, but the workaround to that is simply to `#r` it in both a C# cell and an F# cell. I'll open an issue on your fork if you'd like.

![image](https://github.com/user-attachments/assets/3afb98c3-aca4-45e3-9131-357adf5d9d1b)

Very, very cool!

**Reply by weianweigan** [reply]
@reinux  Thanks! I extended the notebook's functionality using [this tutorial](https://github.com/dotnet/interactive/blob/main/docs/extending-dotnet-interactive.md). I suspect that the notebook generates two separate projects in a temporary directory for C# and F# to support running scripts, which is why they can't share the package. I'd be happy for you to open an issue, but I’m not sure how to fix it.

**Reply by reinux** [reply]
@weianweigan Yeah, I'm not sure either, which is why I've been a bit reluctant. I guess it wouldn't hurt to leave it open for now. Maybe someone will come along who has experience with this.

**Reply by weianweigan** [reply]
Yeah, but I tend to think this might be an issue with the notebook itself.

**Comment by edumarzolla** [comment]
Ok, I'll check it out tomorrow.

**Reply by pca006132** [reply]
Thanks, but I think the issue is now fixed so no need to check now.

**Comment by pca006132** [comment]
Hi, I wonder if you can change the C# binding to an out-of-tree build, and depends on the C bindings. We will be releasing v3.0, we can add the link to your repository.

**Reply by weianweigan** [reply]
Thanks for adding the C# binding link.

The C# binding code is currently in the `/bindings/csharp` folder in my forked repository, and there's also a `nuget.yml` file in the `/.github/workflows` directory. The C# binding depends on the C binding and does not modify the source code of Manifold. I'm not proficient in CMake. What exactly does "out-of-tree build" refer to in this context? Does it mean I need to include the Manifold repository as a submodule?

**Reply by pca006132** [reply]
Oh, by out-of-tree I just mean you can build your binding without modification to the manifold source code, i.e. no extra C bindings needed. It seems that this is the case, so everything is fine :)

**Reply by edumarzolla** [reply]
Dear @weianweigan ,

Regarding https://github.com/elalish/manifold version 3.3.2, I was able to compile the DLL (C Binding) using the following CMAKE command:
cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DMANIFOLD_DEBUG=OFF -DMANIFOLD_PAR=TBB -DMANIFOLD_EXPORT=ON -DMANIFOLD_JSBIND=OFF -DMANIFOLD_PYBIND=OFF -DMANIFOLD_TEST=OFF -DMANIFOLD_DEBUG=OFF -DMANIFOLD_USE_BUILTIN_CLIPPER2=OFF -A x64 -B build
After that, I had to modify the /manifold/cmake/manifoldDeps.cmake file as per the attached file.
[manifoldDeps.zip](https://github.com/user-attachments/files/23917811/manifoldDeps.zip)


This generated the file \manifold\build\bindings\c\manifoldc.vcxproj

Then, after compiling in VS2026, the following files were generated in manifold\build\lib\Release:
manifold.dll
manifold.exp
manifold.lib
manifoldc.dll
manifoldc.exp
manifoldc.lib
tbb12.dll

Based on this explanation, could you please update ManifoldNET.Meshio in your NuGet package?

Best regards,

**Comment by tathuonglong** [comment]
Sorry for resurrect this discussion, but how do I update this binding to the current version of manifold? A few pointers would help a lot. Tried to build this with the instruction within the binding folder but the dll file is way smaller in size and doesn't work.

**Reply by pca006132** [reply]
Can you attach your cmake log here?

**Reply by tathuonglong** [reply]
> @weianweigan [How to compile Manifold DLL in VS2022.docx](https://github.com/user-attachments/files/16736802/How.to.compile.Manifold.DLL.in.VS2022.docx)

Yes, here is the step I took from edumarzolla's instruction, using VS2022 on a Window machine:
- Apply all update from Manifold repo to this fork.
- Run the cmake command in the project root folder: `cmake . -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DMANIFOLD_DEBUG=ON -DMANIFOLD_PAR=TBB -DMANIFOLD_EXPORT=ON -A x64 -B build`
- The log of the last cmake command is as follow:
`E:\Github\manifold-csharp>cmake . -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=OFF -DMANIFOLD_DEBUG=ON -DMANIFOLD_PAR=TBB -DMANIFOLD_EXPORT=OFF -A x64 -B build
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- Clipper2 not found, downloading from source
-- ====================================
-- Manifold Build Configuration Summary
-- ====================================
--
-- Environment: Windows
--
-- CMAKE_VERSION:                 4.2.1
-- CMAKE_TOOLCHAIN_FILE:
-- CMAKE_GENERATOR:               Visual Studio 17 2022
-- CPACK_CMAKE_GENERATOR:
-- CMAKE_BUILD_TYPE:              Release
-- CMAKE_PREFIX_PATH:
-- CMAKE_CXX_COMPILER_ID:         MSVC
-- CMAKE_CXX_COMPILER_VERSION:    19.38.33141.0
-- BUILD_SHARED_LIBS:             OFF
--
-- MANIFOLD_VERSION:              3.3.2
-- MANIFOLD_PAR:                  TBB
-- MANIFOLD_CROSS_SECTION:        ON
-- MANIFOLD_EXPORT:               OFF
-- MANIFOLD_TEST:                 OFF
-- MANIFOLD_DEBUG:                ON
-- MANIFOLD_ASSERT:               OFF
-- MANIFOLD_CBIND:                ON
-- MANIFOLD_PYBIND:               OFF
-- MANIFOLD_JSBIND:               OFF
-- MANIFOLD_FLAGS:                /DNOMINMAX;/bigobj
-- MANIFOLD_COMPILE_OPTIONS:      /DNOMINMAX;/bigobj;-DMANIFOLD_DEBUG;-DMANIFOLD_CROSS_SECTION;-DMANIFOLD_PAR=1
-- MANIFOLD_LINK_OPTIONS:
-- MANIFOLD_DOWNLOADS:            ON
-- MANIFOLD_USE_BUILTIN_TBB:      OFF
-- MANIFOLD_USE_BUILTIN_CLIPPER2: ON
-- MANIFOLD_USE_BUILTIN_NANOBIND: OFF
-- MANIFOLD_FUZZ:                 OFF
-- TRACY_ENABLE:                  OFF
-- TRACY_MEMORY_USAGE:            OFF
--
-- Configuring done (2.1s)
-- Generating done (0.7s)
-- Build files have been written to: E:/Github/manifold-csharp/build`
- Navigate to `manifold-csharp\build\bindings\c` folder, open `manifoldc.vcxproj` as project. Here is the where the instruction is no longer hold up. The instruction said that I should see a "manifold**c**.vcxproj" solution, but I got "manifold.vcxproj" solution instead
- Try to build that solution, the log said that the file "manifold**c**.dll" is in the folder `manifold-csharp\build\lib\Release`. As I said, the library file is really small so I don't know if that is the correct file I'm looking for. Tried to use this file to replace the current dll in the C# binding I got from Nuget. Unfortunately, the file didn't work.

I also try to contact [strich](https://github.com/strich) from discussion #1182 because he was using C# too, but got no response so far.

**Reply by pca006132** [reply]
I think you probably need both the `manifold.dll` and `manifoldc.dll`.

**Reply by tathuonglong** [reply]
So my app just keep crashing without any exception when I put both `manifold.dll` and `manifoldc.dll` into the same folder, so I just clear them all and build a refresh version from this repo, take the freshly `manifold.dll` and `manifoldc.dll` file, put it in my project and it worked, for now. Look likes something was broken in the cache.
